src/HttpServer.gren

```diff
@@ -0,0 +1,303 @@
+        |> Task.andThen (Gren.Kernel.HttpServer.endResponse)
+
+
+-- TODO: tests
+toRequest :
+    { urlProtocol : String
+    , urlHost : String
+    , urlPort : String
+    , urlPath : String
+    , urlQuery : String
+    , urlFragment : String
+    , headers : Array String
+    , method : String
+    , body : String
+    }
+    -> Request
+toRequest 
+    { urlProtocol
+    , urlHost
+    , urlPort
+    , urlPath
+    , urlQuery
+    , urlFragment
+    , headers
+    , method
+    , body 
+    } =
+    { method = toMethod method
+    , body = body
+    , headers =
+        headers
+            |> arrayPairs
+            |> dictFromPairs
+    , url =
+        { protocol = 
+            if urlProtocol == "https:" then
+                Https
+            else
+                Http
+        , host = urlHost
+        , path = urlPath
+        , port_ = String.toInt urlPort
+        , query = 
+            if urlQuery == "" then
+                Nothing
+            else
+                Just urlQuery
+        , fragment =
+            if urlFragment == "" then
+                Nothing
+            else
+                Just urlFragment
+        }
+    }
+
+
+toMethod : String -> Method
+toMethod s =
+    case s of
+        "GET" ->
+            GET
+        
+        "HEAD" ->
+            HEAD
+
+        "POST" ->
+            POST
+        
+        "PUT" ->
+            PUT
+        
+        "DELETE" ->
+            DELETE
+        
+        "CONNECT" ->
+            CONNECT
+        
+        "TRACE" ->
+            TRACE
+        
+        "PATCH" ->
+            PATCH
+        
+        _ ->
+            UNKNOWN
+
+
+arrayPairs : Array String -> Array (Array String)
+arrayPairs a =
+    let
+        pair = 
+            Array.takeFirst 2 a
+        
+        rest =
+            Array.dropFirst 2 a
+        
+        allPairs =
+            [ pair ] ++ case rest of
+                [] ->
+                    []
+                
+                _ ->
+                    arrayPairs rest
+    in
+    allPairs
+
+
+dictFromPairs : Array (Array String) -> Dict String String
+dictFromPairs pairs =
+    let
+        mapper p =
+            case p of
+                [a, b] ->
+                    Just { key = a, value = b }
+                
+                _ ->
+                    Nothing
+        
+        array =
+            Array.filterMap mapper pairs
+    in
+    Dict.fromArray array
+
+
+-- EFFECT STUFF
+
+
+type MySub msg
+    = HandleRequestSub Server (Request -> Response -> msg)
+
+
+subMap : (a -> b) -> MySub a -> MySub b
+subMap f sub =
+    case sub of
+        HandleRequestSub server msg ->
+            -- TODO: not sure this is right
+            HandleRequestSub server (\req res -> f (msg req res))
+
+type alias State msg =
+    { subs : Array (MySub msg)
+    }
+
+
+init : Task Never (State msg)
+init =
+    Task.succeed { subs = [] }
+
+
+-- TODO: Should this task have a real failure type?
```

Comment:
If I remember correctly, you're not allowed to specify a failure type here. The `Task Never a` signature is required by the compiler.